crates/interledger-packet/src/address.rs

```diff
@@ -0,0 +1,385 @@
+                bytes.len(),
+            )));
+        }
+
+        let mut segments = 0;
+        let is_valid_scheme = bytes
+            .split(|&byte| byte == b'.')
+            .enumerate()
+            .all(|(i, segment)| {
+                segments += 1;
+                let scheme_ok = i != 0 || is_scheme(segment);
+                scheme_ok
+                    && !segment.is_empty()
+                    && segment.iter().all(|&byte| is_segment_byte(byte))
+            });
+
+        if is_valid_scheme && segments > 1 {
+            Ok(Address(bytes))
+        } else {
+            Err(ParseError::InvalidAddress(AddressError::InvalidFormat))
+        }
+    }
+}
+
+impl TryFrom<&[u8]> for Address {
+    type Error = ParseError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        Self::try_from(Bytes::from(bytes))
+    }
+}
+
+impl std::ops::Deref for Address {
+    type Target = str;
+
+    fn deref(&self) -> &str {
+        unsafe { str::from_utf8_unchecked(self.0.as_ref()) }
+    }
+}
+
+impl AsRef<[u8]> for Address {
+    #[inline]
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl AsRef<Bytes> for Address {
+    #[inline]
+    fn as_ref(&self) -> &Bytes {
+        &self.0
+    }
+}
+
+impl fmt::Debug for Address {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter
+            .debug_tuple("Address")
+            .field(&self.to_string())
+            .finish()
+    }
+}
+
+impl fmt::Display for Address {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str(self)
+    }
+}
+
+impl Address {
+    /// Returns the length of the ILP Address.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns the `Bytes` conversion of the ILP Address
+    pub fn to_bytes(&self) -> Bytes {
+        self.0.clone()
+    }
+
+    /// Creates an ILP address without validating the bytes.
+    ///
+    /// # Safety
+    ///
+    /// The given bytes must be a valid ILP address.
+    #[inline]
+    pub unsafe fn new_unchecked(bytes: Bytes) -> Self {
+        debug_assert!(Address::try_from(bytes.as_ref()).is_ok());
+        Address(bytes)
+    }
+
+    /// Returns an iterator over all the segments of the ILP Address
+    pub fn segments(&self) -> impl DoubleEndedIterator<Item = &str> {
+        unsafe {
+            self.0
+                .split(|&b| b == b'.')
+                .map(|s| str::from_utf8_unchecked(&s))
+        }
+    }
+
+    /// ```text
+    /// scheme = "g" / "private" / "example" / "peer" / "self" /
+    ///          "test" / "test1" / "test2" / "test3" / "local"
+    /// ```
+    #[inline]
+    pub fn scheme(&self) -> &str {
+        self.segments().next().unwrap()
+    }
+
+    /// Returns the last part (right-most '.' separated segment) of the ILP Address.
+    /// Addresses are guaranteed at least one segment (in addition to the scheme).
+    pub fn last(&self) -> &str {
```

Comment:
```suggestion
    pub fn last_segment(&self) -> &str {
```

It's not clear what `address.last()` would return just from the name.

Also, if this function is only going to be used in a single place in STREAM, I'm not sure it's worth including

Reply:
Yeah `last_segment` is a good idea. 

If there's no point in adding the function, then we can also remove `scheme` and just use the `segments` iterator to access any segment of the address.

Reply:
I think there aren't so many places we need to access it so removing the methods seems fine